File: Sample/SampleOC/SampleAlertViewController.m
//
//  SampleAlertViewController.m
//  SampleOC
//
//  Created by KK on 2026/7/4.
//

#import "SampleAlertViewController.h"

@interface SampleAlertViewController ()
@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) UIButton *closeButton;
@property (nonatomic, strong) NSLayoutConstraint *contentViewHeightLayoutConstraint;
@end

@implementation SampleAlertViewController
#if DEBUG
- (void)dealloc {
    NSLog(@"[%@(%p) dealloc]", NSStringFromClass(self.class), self);
}
#endif

- (void)viewDidLoad {
    [super viewDidLoad];

    self.view.backgroundColor = [UIColor.blackColor colorWithAlphaComponent:0.75];

    [self setupViews];
}

- (void)setupViews {
    _contentView = [UIView new];
    _contentView.backgroundColor = UIColor.whiteColor;
    _contentView.layer.cornerRadius = 12;
    _contentView.translatesAutoresizingMaskIntoConstraints = NO;

    UIButtonConfiguration *configuration = [UIButtonConfiguration plainButtonConfiguration];
    configuration.title = @"关闭";

    _closeButton = [UIButton buttonWithConfiguration:configuration primaryAction:nil];
    _closeButton.translatesAutoresizingMaskIntoConstraints = NO;
    [_closeButton addTarget:self action:@selector(closeButtonAction:) forControlEvents:UIControlEventTouchUpInside];

    [self.view addSubview:_contentView];
    [_contentView addSubview:_closeButton];

    _contentViewHeightLayoutConstraint = [_contentView.heightAnchor constraintEqualToConstant:100];

    [NSLayoutConstraint activateConstraints:@[
        [_contentView.widthAnchor constraintEqualToConstant:320],
        _contentViewHeightLayoutConstraint,
        [_contentView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [_contentView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor],

        [_closeButton.trailingAnchor constraintEqualToAnchor:_contentView.trailingAnchor constant:-20],
        [_closeButton.topAnchor constraintEqualToAnchor:_contentView.topAnchor constant:20],
    ]];
}

- (void)closeButtonAction:(UIButton *)sender {
    !self.didTapCloseHandler ?: self.didTapCloseHandler(self);
}

#pragma mark - OTKOverlayTransitionDelegate
- (id<UIViewImplicitlyAnimating>)overlayTransitionAnimatorForShowingFrom:(UIViewController *)fromViewController to:(UIViewController *)toViewController {

    self.contentViewHeightLayoutConstraint.constant = 200;
    [self.view setNeedsLayout];
    [self.view layoutIfNeeded];

    self.view.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.0];

    self.contentView.hidden = NO;
    self.contentView.transform = CGAffineTransformMakeScale(0.01, 0.01);

    __weak typeof(self) weakSelf = self;
    UIViewPropertyAnimator *animator = [[UIViewPropertyAnimator alloc] initWithDuration:0.25 curve:UIViewAnimationCurveEaseInOut animations:^{
        weakSelf.view.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.7];
        weakSelf.contentView.transform = CGAffineTransformIdentity;
    }];

    return animator;
}

- (id<UIViewImplicitlyAnimating>)overlayTransitionAnimatorForHidingFrom:(UIViewController *)fromViewController to:(UIViewController *)toViewController {
    __weak typeof(self) weakSelf = self;
    UIViewPropertyAnimator *animator = [[UIViewPropertyAnimator alloc] initWithDuration:0.25 curve:UIViewAnimationCurveEaseInOut animations:^{
        weakSelf.view.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.0];
        weakSelf.contentView.transform = CGAffineTransformMakeScale(0.01, 0.01);
    }];

    /* clang-format off */
      [animator addAnimations:^{
          weakSelf.contentView.alpha = 0.0;
      } delayFactor:0.15];
    /* clang-format on */

    return animator;
}

@end

File: Sample/Sample/SampleAlertViewController.swift
//
//  SampleAlertViewController.swift
//  Sample
//
//  Created by KK on 2026/7/4.
//

import OverlayTransitionKit
import UIKit

class SampleAlertViewController: UIViewController, OverlayTransitionDelegate {
    var didTapCloseHandler: ((SampleAlertViewController) -> Void)?

    private var contentView: UIView!
    private var closeButton: UIButton!
    private var contentViewHeightLayoutConstraint: NSLayoutConstraint!

    #if DEBUG
    deinit {
        print("\(self) deinit")
    }
    #endif

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.black.withAlphaComponent(0.75)

        setupViews()
    }

    private func setupViews() {
        contentView = UIView()
        contentView.backgroundColor = .white
        contentView.layer.cornerRadius = 12
        contentView.translatesAutoresizingMaskIntoConstraints = false

        var configuration = UIButton.Configuration.plain()
        configuration.title = "关闭"

        closeButton = UIButton(configuration: configuration, primaryAction: nil)
        closeButton.translatesAutoresizingMaskIntoConstraints = false
        closeButton.addTarget(self, action: #selector(closeButtonAction(_:)), for: .touchUpInside)

        view.addSubview(contentView)
        contentView.addSubview(closeButton)

        contentViewHeightLayoutConstraint = contentView.heightAnchor.constraint(equalToConstant: 100)

        NSLayoutConstraint.activate([
            contentView.widthAnchor.constraint(equalToConstant: 320),
            contentViewHeightLayoutConstraint,
            contentView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            contentView.centerYAnchor.constraint(equalTo: view.centerYAnchor),

            closeButton.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -20),
            closeButton.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 20),
        ])
    }

    @objc
    private func closeButtonAction(_ sender: UIButton) {
        didTapCloseHandler?(self)
    }

    // MARK: - OverlayTransitionDelegate

    func overlayTransitionAnimatorForShowing(from fromViewController: UIViewController, to toViewController: UIViewController) -> UIViewImplicitlyAnimating {
        contentViewHeightLayoutConstraint.constant = 200
        view.setNeedsLayout()
        view.layoutIfNeeded()

        view.backgroundColor = UIColor.black.withAlphaComponent(0.0)

        contentView.isHidden = false
        contentView.transform = CGAffineTransform(scaleX: 0.01, y: 0.01)

        return UIViewPropertyAnimator(duration: 0.25, curve: .easeInOut) { [weak self] in
            self?.view.backgroundColor = UIColor.black.withAlphaComponent(0.7)
            self?.contentView.transform = .identity
        }
    }

    func overlayTransitionAnimatorForHiding(from fromViewController: UIViewController, to toViewController: UIViewController) -> UIViewImplicitlyAnimating {
        let animator = UIViewPropertyAnimator(duration: 0.25, curve: .easeInOut) { [weak self] in
            self?.view.backgroundColor = UIColor.black.withAlphaComponent(0.0)
            self?.contentView.transform = CGAffineTransform(scaleX: 0.01, y: 0.01)
        }

        animator.addAnimations({ [weak self] in
            self?.contentView.alpha = 0.0
        }, delayFactor: 0.15)

        return animator
    }
}
